File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def in_range(x, y, n, m):
    return 0 <= x < n and 0 <= y < m

def bfs(maps):
    n = len(maps)
    m = len(maps[0])
    
    dxs = [1, 0, -1, 0]
    dys = [0, 1, 0, -1]
    visited = [[False]*m for _ in range(n)]
    
    visited[0][0] = True
    queue = deque([(0, 0, 1)])
    
    while queue:
        x, y, s = queue.popleft()
        if x == n-1 and y == m-1:
            return s
        for dx, dy in zip(dxs, dys):
            nx, ny = x+dx, y+dy
            if in_range(nx, ny, n, m) and maps[nx][ny] == 1 and not visited[nx][ny]:
                visited[nx][ny] = True
                queue.append((nx, ny, s+1))
    return -1

def solution(maps):
    step = bfs(maps)
    return -1 if step == -1 else step

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;


class Solution {
    
    static int[] dx = new int[]{-1, 0, 1, 0};
    static int[] dy = new int[]{0, 1, 0, -1};
    static int n;
    static int m;
    
    public int solution(int[][] maps) {
        int answer = 0;
        n = maps.length;
        m = maps[0].length;
        
        boolean[][] visited = new boolean[n][m];
        Queue<int[]> queue = new LinkedList<>();
        
        queue.offer(new int[]{0, 0, 1});
        while (!queue.isEmpty()) {
            int[] p = queue.poll();
            int x = p[0], y = p[1];
            int cnt = p[2];
            
            for (int i = 0; i < 4; i++) {
                int nx = x+dx[i];
                int ny = y+dy[i];
                
                if (nx == (n-1) && ny == (m-1)) {
                    return cnt+1;
                }
                
                if (inRange(nx, ny) && maps[nx][ny] == 1 && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    queue.offer(new int[]{nx, ny, cnt+1});
                }
            }
        }
        
        return -1;
    }
    
    public boolean inRange(int x, int y) {
        return 0 <= x && x < n && 0 <= y && y < m;
    }
}